src/multiselect.js

```diff
@@ -0,0 +1,206 @@
+ * Class for using multiple select blocks on workspace.
+ */
+export class Multiselect {
+  /**
+   * Initalize the class data structure.
+   * @param {!Blockly.WorkspaceSvg} workspace The workspace to sit in.
+   */
+  constructor(workspace) {
+    this.workspace_ = workspace;
+    this.origHandleWsStart_ = Blockly.Gesture.prototype.handleWsStart;
+  }
+
+  /**
+   * Bind the events and replace registration.
+   * @param {{multiselectIcon: Object, useDoubleClick: boolean}} options
+   * to set.
+   */
+  init(options) {
+    this.onKeyDownWrapper_ = Blockly.browserEvents.conditionalBind(
+        this.workspace_.getInjectionDiv(), 'keydown', this, this.onKeyDown_);
+    this.onKeyUpWrapper_ = Blockly.browserEvents.conditionalBind(
+        this.workspace_.getInjectionDiv(), 'keyup', this, this.onKeyUp_);
+    this.eventListenerWrapper_ = this.eventListener_.bind(this);
+    this.workspace_.addChangeListener(this.eventListenerWrapper_);
+
+    ContextMenu.unregisterContextMenu();
+    ContextMenu.registerOurContextMenu();
+    Shortcut.unregisterShortcut();
+    Shortcut.registerOurShortcut();
+
+    this.controls_ = new MultiselectControls(
+        this.workspace_, options.multiselectIcon, this);
+    if (!options.multiselectIcon || !options.multiselectIcon.hideIcon) {
+      const svgControls = this.controls_.createDom();
+      this.workspace_.getParentSvg().appendChild(svgControls);
+    }
+    this.controls_.init();
+
+    if (options.useDoubleClick) {
+      this.useDoubleClick_(true);
+    }
+
+    if (options.baseBlockDragger) {
+      setBaseBlockDragger(options.baseBlockDragger);
+    }
+
+    if (!options.bumpNeighbours) {
+      this.origBumpNeighbours = Blockly.BlockSvg.prototype.bumpNeighbours;
+      Blockly.BlockSvg.prototype.bumpNeighbours = function() {};
+    }
+  }
+
+  /**
+   * Unbind the events and replace with original registration.
+   */
+  dispose() {
+    if (this.onKeyDownWrapper_) {
+      Blockly.browserEvents.unbind(this.onKeyDownWrapper_);
+      this.onKeyDownWrapper_ = null;
+    }
+    if (this.onKeyUpWrapper_) {
+      Blockly.browserEvents.unbind(this.onKeyUpWrapper_);
+      this.onKeyUpWrapper_ = null;
+    }
+    if (this.eventListenerWrapper_) {
+      this.workspace_.removeChangeListener(this.eventListenerWrapper_);
+      this.eventListenerWrapper_ = null;
+    }
+
+    ContextMenu.unregisterContextMenu();
+    ContextMenu.registerOrigContextMenu();
+    Blockly.ContextMenuRegistry.registry.unregister('workspaceSelectAll');
+    Shortcut.unregisterShortcut();
+    Blockly.ShortcutRegistry.registry.unregister('selectall');
+    Shortcut.registerOrigShortcut();
+
+    if (this.controls_) {
+      this.controls_.dispose();
+      this.controls_ = null;
+    }
+
+    this.useDoubleClick_(false);
+
+    if (this.origBumpNeighbours) {
+      Blockly.BlockSvg.prototype.bumpNeighbours = this.origBumpNeighbours;
+    }
+  }
+
+  /**
+   * Add double click to expand/collapse blocks (MIT App Inventor Supported).
```

Comment:
If time permits, we may want to break this out as a separate plugin.

Reply:
Okay, I see. Shall we remove the feature in this plugin before release or keep it as for now?

Reply:
Okay, so let's keep it for now.